ext/http/http_next.rs

```diff
@@ -0,0 +1,800 @@
+      request_info,
+      request_parts,
+      request_body: Some(request_body),
+      response: Some(Response::new(ResponseBytes::default())),
+      body: None,
+      promise: CompletionHandle::default(),
+      #[cfg(zombie_http_tracking)]
+      alive: true,
+    })
+  })
+}
+
+struct SafeFutureForSingleThread<F: Future<Output = O> + Unpin, O>(F);
+
+impl<F: Future<Output = O> + Unpin, O> Future
+  for SafeFutureForSingleThread<F, O>
+{
+  type Output = O;
+  fn poll(
+    mut self: Pin<&mut Self>,
+    cx: &mut std::task::Context<'_>,
+  ) -> std::task::Poll<Self::Output> {
+    self.0.poll_unpin(cx)
+  }
+}
+
+// SAFETY: We use this to allow us to use tokio::task::spawn in a single-threaded executor.
+unsafe impl<F: Future<Output = O> + Unpin, O> Send
+  for SafeFutureForSingleThread<F, O>
+{
+}
+
+#[op]
+pub fn op_upgrade_raw(_index: usize) {}
+
+#[op]
+pub async fn op_upgrade(
+  state: Rc<RefCell<OpState>>,
+  index: usize,
+  headers: Vec<(ByteString, ByteString)>,
+) -> Result<ResourceId, AnyError> {
+  // Stage 1: set the respnse to 101 Switching Protocols and send it
+  let upgrade = with_http_mut(index, |http| {
+    // Manually perform the upgrade. We're peeking into hyper's underlying machinery here a bit
+    let upgrade = http.request_parts.extensions.remove::<OnUpgrade>().unwrap();
+
+    let response = http.response.as_mut().unwrap();
+    *response.status_mut() = StatusCode::SWITCHING_PROTOCOLS;
+    for (name, value) in headers {
+      response.headers_mut().append(
+        HeaderName::from_bytes(&name).unwrap(),
+        HeaderValue::from_bytes(&value).unwrap(),
+      );
+    }
+    http.promise.complete(true);
+    upgrade
+  });
+
+  // Stage 2: wait for the request to finish upgrading
+  let upgraded = upgrade.await?;
+
+  // Stage 3: return the extracted raw network stream
+  let upgraded = match upgraded.downcast::<tokio::net::TcpStream>() {
+    Ok(hyper1::upgrade::Parts { io: conn, .. }) => {
+      return put_network_stream_resource(
+        &mut state.borrow_mut().resource_table,
+        NetworkStream::Tcp(conn),
+      );
+    }
+    Err(x) => x,
+  };
+  let upgraded = match upgraded.downcast::<tokio::net::UnixStream>() {
+    Ok(hyper1::upgrade::Parts { io: conn, .. }) => {
+      return put_network_stream_resource(
+        &mut state.borrow_mut().resource_table,
+        NetworkStream::Unix(conn),
+      );
+    }
+    Err(x) => x,
+  };
+  let _upgraded = match upgraded.downcast::<deno_net::ops_tls::TlsStream>() {
+    Ok(hyper1::upgrade::Parts { io: conn, .. }) => {
+      return put_network_stream_resource(
+        &mut state.borrow_mut().resource_table,
+        NetworkStream::Tls(conn),
+      );
+    }
+    Err(x) => x,
+  };
+  Err(bad_resource("Impossible to upgrade this connection"))
+}
+
+#[op]
+pub fn op_set_promise_complete(index: usize, status: u16) {
+  //println!("promise...");
+  with_resp_mut(index, |resp| {
+    // TODO(mmastrac): Fail if invalid status code
```

Comment:
Can you address this TODO?

Reply:
Replaced w/`unwrap` as the JS code cannot actually send an invalid value.